### PR TITLE
Dependency graph for task handling

### PR DESCRIPTION
**Pull request by willianrampazzo**
This introduces a new architecture responsible for handling the tasks dependencies.

Signed-off-by: Willian Rampazzo <willianr@redhat.com>

Changes from the RFC:

- improved stages description;
- clarify what the dependency graph does not solve;
- improve text in the DOT section;
- use a valid DOT example;
- added the BP004 to the index;
- minor text fixes.

### CODE REVIEW

File: docs/source/blueprints/BP004.rst
Comment (clebergnu) [inline docs/source/blueprints/BP004.rst:206]: If we have an issue about the tests that are being skipped and that would need to be enabled once this is implemented, let's add it here.

For comparison sake, BP002 had a number of issues listed here.  I'm not sure how much of the work can be foreseen and have issues created here, but if you and @richtja have an idea of the next steps it would be a good idea to add them here as issues too.

**Review by clebergnu — APPROVED** [review]
Other than the comment about adding issues (if applicable) this LGTM.

File: docs/source/blueprints/BP004.rst
Comment (willianrampazzo) [reply docs/source/blueprints/BP004.rst:206]: > If we have an issue about the tests that are being skipped and that would need to be enabled once this is implemented, let's add it here.

Okay, I will do that.

> 
> For comparison sake, BP002 had a number of issues listed here. I'm not sure how much of the work can be foreseen and have issues created here, but if you and @richtja have an idea of the next steps it would be a good idea to add them here as issues too.

On BP002, we had a clearer view of what could become an issue. I opened and added all of them right after the blueprint was approved. I'll discuss with @richtja which issues we can open and add them to the blueprint as soon as the PR has all the approvals needed.

Comment (willianrampazzo) [reply docs/source/blueprints/BP004.rst]: I force-pushed with the issue that is already opened about the tasks clash.

**Review by richtja — APPROVED** [review]
LGTM, thanks

File: docs/source/blueprints/BP004.rst
Comment (richtja) [reply docs/source/blueprints/BP004.rst]: > 
>On BP002, we had a clearer view of what could become an issue. I opened and added all of them right after the >blueprint was approved. I'll discuss with @richtja which issues we can open and add them to the blueprint as soon as >the PR has all the approvals needed.

Yes, when this will be approved, we can discuss and create epic issue to track all the tasks which have to be done.

File: docs/source/blueprints/BP004.rst
Comment (beraldoleal) [inline docs/source/blueprints/BP004.rst:8]: Maybe you could add the reviews at this point.

File: docs/source/blueprints/BP004.rst
Comment (beraldoleal) [inline docs/source/blueprints/BP004.rst:63]: nrunner runner ? Maybe nrunner_runner ?

**Review by beraldoleal — APPROVED** [review]
This LGTM, just two small comments.

File: docs/source/blueprints/BP004.rst
Comment (willianrampazzo) [reply docs/source/blueprints/BP004.rst:8]: Sure, I can do that when I add the issues.

File: docs/source/blueprints/BP004.rst
Comment (willianrampazzo) [reply docs/source/blueprints/BP004.rst:63]: In the first sentence, I'm naming the plugin, not the file; that is why I added a space between nrunner and runner. I understand that it is called `nrunner_runner` because we do not want to add spaces on a file name.

**Comment by willianrampazzo** [comment]
I changed the status to `approved` and added the reviewers. I'll add it as soon as we have the epic issue, and the PR can be merged.

File: docs/source/blueprints/BP004.rst
Comment (beraldoleal) [reply docs/source/blueprints/BP004.rst:63]: Based on our previous documents, and the plugin name itself I find confusing. Specially because is using the "code" syntax.

Comment (willianrampazzo) [reply docs/source/blueprints/BP004.rst:63]: Could you point me to a document where it calls a plugin like `some_plugin`? I could not find any occurrence on the documentation. I wanted to highlight the term, but I can remove it and let it be in plain text if that makes the text less confusing, no problem.

Comment (beraldoleal) [reply docs/source/blueprints/BP004.rst:63]: Btw, not a blocker, just a comment. I could not find reference to "nrunner runner" plugin there:

 `grep plugin docs/* -R | grep \` | grep runner`.
 
 Looks like we do have different forms there. But my point is that the plugin is called "runner_nrunner.py" I could not find "nrunner_runner.py".

**Comment by willianrampazzo** [comment]
I added an Epic issue to the list of issues. From what we discussed, the PR is ready to be merged. Let me know if there is something else needed before it gets merged.

**Pull request merged**
